packages/block-editor/src/store/reducer.js

```diff
@@ -187,15 +187,30 @@ export function isUpdatingSameBlockAttribute( action, lastAction ) {
 function withPersistentBlockChange( reducer ) {
 	let lastAction;
 
+	/**
+	 * Set of action types for which a blocks state change should be considered
+	 * non-persistent.
+	 *
+	 * @type {Set}
+	 */
+	const IGNORED_ACTION_TYPES = new Set( [
+		'RECEIVE_BLOCKS',
+	] );
+
 	return ( state, action ) => {
 		let nextState = reducer( state, action );
+
 		const isExplicitPersistentChange = action.type === 'MARK_LAST_CHANGE_AS_PERSISTENT';
 
 		if ( state !== nextState || isExplicitPersistentChange ) {
+			// Some state changes should not be considered persistent, namely
+			// those which are not a direct result of user interaction.
+			const isPersistentStateChange = ! IGNORED_ACTION_TYPES.has( action.type );
```

Comment:
The change seems logical but it doesn't seem more stable. I'm thinking maybe we shouldn't update  `lastAction` if the action is ignored.

Reply:
> The change seems logical but it doesn't seem more stable. I'm thinking maybe we shouldn't update `lastAction` if the action is ignored.

Yes, based on [the equivalent implementation](https://github.com/WordPress/gutenberg/blob/cd82d07a70c5485396427f7196e78ae5c2d3b284/packages/editor/src/utils/with-history/index.js#L120-L137) from `withHistory`, that might be the better option.

Generally, this all feels quite unstable, but always has. I think it's worth making a dedicated push to refactor reusable blocks.